src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/UpToDateCheckImplicitConfiguredInput.cs

```diff
@@ -0,0 +1,572 @@
+                        {
+                            itemsBySet[setName] = builder = ImmutableHashSet.CreateBuilder(BuildUpToDateCheck.SetNameComparer);
+                        }
+
+                        builder.Add(destination);
+                    }
+                }
+
+                upToDateCheckBuiltItems = itemsBySet.ToImmutableDictionary(pair => pair.Key, pair => pair.Value.ToImmutable(), BuildUpToDateCheck.SetNameComparer);
+                copiedOutputFiles = copiedOutputFilesBuilder.ToImmutable();
+            }
+            else
+            {
+                upToDateCheckBuiltItems = UpToDateCheckBuiltItemsBySetName;
+                copiedOutputFiles = CopiedOutputFiles;
+            }
+
+            var itemTypes = projectItemSchema
+                .GetKnownItemTypes()
+                .Where(itemType => projectItemSchema.GetItemType(itemType).UpToDateCheckInput)
+                .ToImmutableHashSet(StringComparers.ItemTypes);
+
+            var itemTypeDiff = new SetDiff<string>(ItemTypes, itemTypes, StringComparers.ItemTypes);
+
+            var itemsByItemTypeBuilder = ItemsByItemType.ToBuilder();
+
+            bool itemTypesChanged = false;
+
+            List<(bool IsAdd, string ItemType, string Path, string? Link, BuildUpToDateCheck.CopyType)> changes = new();
+
+            foreach (string removedItemType in itemTypeDiff.Removed)
+            {
+                itemTypesChanged = true;
+
+                if (itemsByItemTypeBuilder.TryGetValue(removedItemType, out var removedItems))
+                {
+                    foreach ((string path, string? link, BuildUpToDateCheck.CopyType copyType) in removedItems)
+                    {
+                        changes.Add((false, removedItemType, path, link, copyType));
+                    }
+
+                    itemsByItemTypeBuilder.Remove(removedItemType);
+                }
+            }
+
+            itemTypesChanged |= itemTypeDiff.Added.GetEnumerator().MoveNext();
+
+            bool itemsChanged = false;
+
+            foreach ((string schemaName, IProjectChangeDescription projectChange) in sourceItemsUpdate.ProjectChanges)
+            {
+                // ProjectChanges is keyed by the rule name which is usually the same as the item type, but not always (eg, in auto-generated rules)
+                string? itemType = null;
+                if (projectCatalogSnapshot.NamedCatalogs.TryGetValue(PropertyPageContexts.File, out IPropertyPagesCatalog fileCatalog))
+                    itemType = fileCatalog.GetSchema(schemaName)?.DataSource.ItemType;
+
+                if (itemType == null)
+                    continue;
+                if (!itemTypes.Contains(itemType))
+                    continue;
+                if (!itemTypesChanged && !projectChange.Difference.AnyChanges)
+                    continue;
+                if (projectChange.After.Items.Count == 0)
+                    continue;
```

Comment:
`or` all of these conditions together into 1 expression?
```suggestion
                if(itemType == null ||
                   !itemTypes.Contains(itemType) ||
                   (!itemTypesChanged && !projectChange.Difference.AnyChanges) ||
                   projectChange.After.Items.Count == 0)
                {
                  continue;
                }
```

Reply:
I agree they're logically equivalent. I find the earlier form easier to read.